docs/control-structures.rst

```diff
@@ -625,21 +627,24 @@ The error code supplied with the error data indicates the kind of panic.
 #. 0x41: If you allocate too much memory or create an array that is too large.
 #. 0x51: If you call a zero-initialized variable of internal function type.
 
-The ``require`` function either creates an error without any data or
-an error of type ``Error(string)``. It
-should be used to ensure valid conditions
-that cannot be detected until execution time.
-This includes conditions on inputs
-or return values from calls to external contracts.
+The ``require`` function provides three overloads:
+
+1. ``require(bool)`` which will revert without any data (not even an error selector).
+2. ``require(bool, string)`` which will revert with an ``Error(string)``.
+3. ``require(bool, error)`` which will revert with the custom, user supplied error provided as the second argument.
 
 .. note::
+    ``require`` arguments are evaluated unconditionally, so take special care to make sure that
+    they are not expressions with unexpected side-effects.
+    For example, in ``require(condition, CustomError(f()));`` and ``require(condition, f());``,
+    function ``f()`` will be called regardless of whether the supplied condition is ``true`` or ``false``.
 
-    It is currently not possible to use custom errors in combination
-    with ``require``. Please use ``if (!condition) revert CustomError();`` instead.
+.. note::
+    Using custom errors with ``require`` is only supported by the via IR` pipeline, i.e. compilation via Yul.
```

Comment:
```suggestion
    Using custom errors with ``require`` is only supported by the via IR pipeline, i.e. compilation via Yul.
```

@cameel SHAME

Reply:
What? This bit has the same awkward wording, but it wasn't me that introduced it: https://github.com/ethereum/solidity/commit/c48389ade2cf99df3fd9c705b88a0dac1d1eff30#diff-228c7a9db0c2e0fe613d1becd9444da3c16bfdb33455cdfaa825c187399f61fdL642-R643

I saw it but ignored it back then. TBH I was most bothered by it being used in a script, which meant I could not avoid using it myself. What I wanted to push back against was enshrining it as *the* way to spell it.

Anyway, this turned into a weird exchange. This is now fixed and as far as I can tell https://github.com/ethereum/solidity/pull/14913#discussion_r1576445073 is the only outstanding comment. Add the comment, rename the template variable and we're done here.

Reply:
Hahaha, it's a joke comment Kamil, I was surprised you missed the superfluous backtick 😁

Reply:
Ah, that's what you meant :)

Sorry, I get a bit nervous these days when an exchange on a small detail that should have been just "fix and forget" gets too drawn out, because I don't want it to turn into something that stalls the PR :)